src/builder.rs

```diff
@@ -768,7 +768,7 @@ fn prepare_header_path(dst: &mut dyn Write, header: &mut Header, path: &Path) ->
             Ok(s) => s,
             Err(e) => str::from_utf8(&data[..e.valid_up_to()]).unwrap(),
         };
-        header.set_path(truncated)?;
+        header.set_truncated_path_for_gnu_header(&truncated)?;
```

Comment:
> but what I feel would improve things here is if we did a cast `as_gnu_mut()` and operated on that instead or so

I agree that it's a good idea but unfortunately a major refactor. From what I can see the problem starts at: [`prepare_header_path`](https://github.com/alexcrichton/tar-rs/blob/main/src/builder.rs#L744) function which acts as de-facto `set_path` that can handle GNU header. I think it's done this way because unlike other `set_path` functions (the vanilla and ustar one), this one doesn't just influence header structure but emits a new entry into archive containing that long path.